File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        
         int n = nums.length;
        TreeMap<Integer, Integer> mp = new TreeMap<>();
        int[] ans = new int[n - k + 1];
        
       
        for (int i = 0; i < k; i++) {
            mp.put(nums[i], mp.getOrDefault(nums[i], 0) + 1);
        }
        
       
        ans[0] = mp.lastKey();
        
        
        for (int i = k; i < n; i++) {
        
            int prev = mp.get(nums[i - k]);
            if (prev == 1) {
                mp.remove(nums[i - k]);
            } else {
                mp.put(nums[i - k], prev - 1);
            }
            
        
            mp.put(nums[i], mp.getOrDefault(nums[i], 0) + 1);
            
        
            ans[i - k + 1] = mp.lastKey();
        }
        
        return ans;
        
        
        
        
    }
}

File: 2025-and-before/0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    
    void helper(deque<array<int,2>>&dq,int num){
        
        while(dq.size() and dq.front()[0]<=num)
            dq.pop_front();
    }
    
    
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        
        vector<int>ans;
        deque<array<int,2>>dq;
        
        for(int i=0;i<k;i++){
            helper(dq,nums[i]);
            dq.push_front({nums[i],i});
        }
        
        ans.push_back(dq.back()[0]);
        
        for(int i=k;i<nums.size();i++){
            
            while(dq.size() and dq.back()[1]<=i-k)
                dq.pop_back();
            
            helper(dq,nums[i]);
            dq.push_front({nums[i],i});
            
            ans.push_back(dq.back()[0]);
        }
        
        return ans;
        
    }
};